src/GrantReservations.sol

```diff
@@ -0,0 +1,187 @@
+    /// @param timestamp The timestamp of the reservation
+    /// @param receiver The address that will receive the tokens (this is also the signal of the ZKP)
+    /// @param root The root of the Merkle tree (signup-sequencer or world-id-contracts provides this)
+    /// @param nullifierHash The nullifier for this proof, preventing double signaling
+    /// @param proof The zero knowledge proof that demonstrates the claimer has a verified World ID
+    /// @dev hashToField function docs are in lib/world-id-contracts/src/libraries/ByteHasher.sol
+    function claim(uint256 timestamp, address receiver, uint256 root, uint256 nullifierHash, uint256[8] calldata proof, bytes calldata signature)
+        external
+    {
+
+        IGrant grant = recurringGrantDrop.grant();
+        uint256 grantId = grant.calculateId(timestamp);
+        checkClaim(timestamp, receiver, root, nullifierHash, proof, signature);
+
+        nullifierHashes[nullifierHash] = true;
+
+        SafeERC20.safeTransferFrom(recurringGrantDrop.token(), recurringGrantDrop.holder(), receiver, grant.getAmount(grantId));
+
+        emit GrantClaimed(grantId, receiver);
+    }
+
+    /// @notice Check whether a claim is valid
+    /// @param timestamp The timestamp of the reservation
+    /// @param receiver The address that will receive the tokens (this is also the signal of the ZKP)
+    /// @param root The root of the Merkle tree (signup-sequencer or world-id-contracts provides this)
+    /// @param nullifierHash The nullifier for this proof, preventing double signaling
+    /// @param proof The zero knowledge proof that demonstrates the claimer has a verified World ID
+    function checkClaim(uint256 timestamp, address receiver, uint256 root, uint256 nullifierHash, uint256[8] calldata proof, bytes calldata signature)
+        public
+    {
+        if (nullifierHashes[nullifierHash]) revert InvalidNullifier();
+        if (receiver == address(0)) revert InvalidReceiver();
+
+        IGrant grant = recurringGrantDrop.grant();
+        uint256 grantId = grant.calculateId(timestamp);
+        if (grantId == grant.getCurrentId()) revert InvalidGrant();
+
+        // ecrecovery from sigature and check against allowed signers
+        address signer = ECDSA.recover(keccak256(abi.encode(timestamp, nullifierHash)), signature);
+        if (!allowedSigners[signer]) revert UnauthorizedSigner();
+
+        // Try to call check claim on recurringGrantDrop
+        try recurringGrantDrop.checkClaim(grantId, receiver, root, nullifierHash, proof) {
+            // This should not happen since only claiming the current grant can succeed, which is already prohibited by the check above.
+        } catch (bytes memory reason) {
+            // Check if nullifier has already been used.
+            if (bytes4(reason) == RecurringGrantDrop.InvalidNullifier.selector) {
+                revert InvalidNullifier();
+            }
+            // Any other error is fine.
+        }
```

Comment:
Let's abstract this hack to its own function for clarity.

Also, let's think if there's a better way. Maybe redeploy grant with a view function for it.

Reply:
Storage proofs.